Readme for the Testing - detailed instructions for functional testing usage and addition of new tests (#983)
cisagov/ScubaGear: Testing/Readme.md

Reviewer: Recommend putting all documentation (other than the base readme) in one folder (with subfolders as necessary), perhaps called `docs` or `documentation`.

Author: Document structure has been discussed earlier and decided to keep this file under Testing folder for now.

    - [Adding new functional test - Example #2](#adding-new-functional-test---example-2)
  - [Nightly Functional Tests](#nightly-functional-tests)
  - [Troubleshooting](#troubleshooting)
    - [Chrome browser version issue](#chrome-browser-version-issue)
    - [Service principal authentication issue](#service-principal-authentication-issue)

## Functional Testing Prerequisites ##

Running ScubaGear functional test automation require Windows compute or VM end system. This development system should be setup with Pester, Selenium, Chrome and PowerShell 5.1. The repository provides a few utility scripts to install and update these prerequisite components.

### Windows System with Chrome ###

Functional testing needs to be run on a Windows system with Chrome browser installed. Automation suite uses Selenium tool which requires Chrome driver for its web component testing.

### Pester ###

"Pester is a testing and mocking framework for PowerShell." ([Pester Quick Start](https://pester.dev/docs/quick-start))

On your Windows development system, install an updated Pester module by running following command as an administrator:

```

> Install-Module -Name Pester -Force -SkipPublisherCheck

Reviewer: When I ran this command, I got the following warning:

`WARNING: The version '5.5.0' of module 'Pester' is currently in use. Retry the operation after closing the
applications.`

Please add this warning to the instructions and include instructions for how to respond to it.

Author: Resolved with an additional note in the troubleshooting section.

> Install-Module -Name Pester -Force -SkipPublisherCheck

Author: PowerShell version is same as ScubaGear and is mentioned in the first few lines of prereq section.

> Install-Module -Name Pester -Force -SkipPublisherCheck

Author: Instll-Module will install latest version of Pester. We are not (yet) specifying required version number, assumption is latest version of Pester/Selenium works. If there are issues in future, we need to update this. The troubleshooting section has info on how to uninstall previous versions before installing the latest version.

> Install-Module -Name Pester -Force -SkipPublisherCheck

Reviewer 2: FYI - on some Windows systems, a very old version of Pester comes pre-installed. I can verify that running with the new version even though the old version is on the machine seems to work okay with our functional tests. I have seen this on a few different Windows machines.

![image](https://github.com/cisagov/ScubaGear/assets/107076927/43118418-cc61-4f8d-9112-7d3f30163681)

- [ ] Ensure all Non-Complaint configuration options for the affected policy are tested.
- [ ] Ensure all Complaint configuration options for the affected policy are tested.
- [ ] Ensure that any changes to HTML report output are tested.
- [ ] Validate that all new functional tests pass on CI for the branch before creating the PR

### Adding new functional test - Example #1 ###

Use case: functional test for a SCB where there is a single non-complaint configuration option and a single complaint configuration option.

Example SCB: MS.POWERPLATFORM.1.1v1 - the ability to create production and sandbox environments should be restricted to admin users

Configuration options: looking into the tenant, there are only two configurations options for restricting the ability to create production and sandbox environments. Non-complaint config: allowing everyone to create production and sandbox environments. Complaint config: restricting it only specific admins.

![Power Platform 1.1 Config](/images/PP-1-1.png)

> Automated functional test case for non-complaint and complaint configs:

Reviewer: ```suggestion
This is an automated functional test case for both non-compliant and compliant configs:
```

```
  - PolicyId: MS.POWERPLATFORM.1.1v1
    TestDriver: RunScuba
    Tests:
      - TestDescription: MS.POWERPLATFORM.1.1v1 Non-Compliant case - Allows production/sandbox environment creation by nonadmin users
        Preconditions:
          - Command: Set-TenantSettings
            Splat:
              RequestBody:
                DisableEnvironmentCreationByNonAdminUsers: false
        Postconditions: []
        ExpectedResult: false
      - TestDescription: MS.POWERPLATFORM.1.1v1 Compliant case - Restricts production/sandbox environment creation to admin users
        Preconditions:
          - Command: Set-TenantSettings
            Splat:
              RequestBody:
                DisableEnvironmentCreationByNonAdminUsers: true
        Postconditions: []
        ExpectedResult: true
```

### Adding new functional test - Example #2 ### 

> Use case: functional test for a SCB where there are multiple non-complaint configuration options.

Reviewer: ```suggestion
**Use case**: Functional test for a SCB where there are multiple non-compliant configuration options.
```

> Example SCB: MS.TEAMS.1.3v1: restricting dial-in users and anonymous users from being admitted into a Teams meeting automatically.

Reviewer: ```suggestion
**Example SCB**: MS.TEAMS.1.3v1: Restricting dial-in users and anonymous users from being admitted into a Teams meeting automatically.
```

> Configuration options: looking into the tenant there are multiple options in allowing and restricting different groups of users to automatically admitted into a meeting. There are two config options that would fail the SCB and others that would pass the SCB.

Reviewer: ```suggestion
**Configuration options**: Looking into the tenant, there are multiple options for allowing and restricting different groups of users to be automatically admitted into a meeting. There are two config options that would fail the SCB, and others that would pass the SCB.
```

![Teams 1.3 Config](/images/Teams-1-3.png)

> Automated functional test case for non-complaint and complaint config options:

Reviewer: ```suggestion
This is an automated functional test case for non-compliant and compliant config options:
```

          - Command: Set-CsTeamsMeetingPolicy
            Splat:
              Identity: Global
              AutoAdmittedUsers: EveryOne
        Postconditions: []
        ExpectedResult: false
      - TestDescription: MS.TEAMS.1.3v1 Non-complaint - Dialin bypass lobby; Everyone not auto admitted user
        Preconditions:
          - Command: Set-CsTeamsMeetingPolicy
            Splat:
              Identity: Global
              AllowPSTNUsersToBypassLobby: true
          - Command: Set-CsTeamsMeetingPolicy
            Splat:
              AutoAdmittedUsers: EveryoneInCompany
              Identity: Global
        Postconditions: []
        ExpectedResult: false
      - TestDescription: MS.TEAMS.1.3v1 Compliant - Dialin not bypass lobby Everyone not auto admitted user
        Preconditions:
          - Command: Set-CsTeamsMeetingPolicy
            Splat:
              Identity: Global
              AllowPSTNUsersToBypassLobby: false
          - Command: Set-CsTeamsMeetingPolicy
            Splat:
              AutoAdmittedUsers: EveryoneInCompany
              Identity: Global
        Postconditions: []
        ExpectedResult: true
```

## Nightly Functional Tests ##

> Product functional tests are being run for multiple product-tenant combinations. Development team should check the successful completion of these runs for any regressions. Currently the following "product / tenant / other variant" combinations are run by a cron job:

Reviewer: ```suggestion
A GitHub workflow is used to execute multiple product functional tests every night, and they are being run for multiple product-tenant combinations. Development teams should check for the successful completion of these runs each morning to ensure that there are not any regressions. As of May 2024, the following "product / tenant / other variant" combinations are run each night:
```

- AAD - G3
- AAD - G5
- AAD - E5
- Defender - G3
- Defender - G5
- EXO
- Power Platform
- SharePoint
- Teams

> 

Reviewer: ```suggestion

The current list of tests is found in a Slack canvas.

```

## Troubleshooting ##

### Chrome browser version issue ###

If the Chrome browser on your system has been updated since you setup the Selenium environment, you may receive an error message like below:

![Chrome browser version](/images/chrome-version.png)

To resolve this, first kill any ChromeDriver.exe processes running by using the Task Manager. If you do not successfully kill that process you might receive an error message such as the one below.

![Chrome browser processes](/images/chrome-process-kill.png)

Then execute the commands below to update the Selenium Chrome driver.

> ```

Reviewer: ```suggestion
```powershell
```

.\Testing\Functional\SmokeTest\UpdateSelenium.ps1
```

> Sometimes the UpdateSelenium script has a problem getting rid of an older chromedriver.exe even if you think you have killed any running processes named chromedriver.exe. You might receive the error "Cannot remove item. Access to the path ...\chromedriver.exe is denied" as in the screenshot above. In that case you should follow the steps below:

Reviewer: ```suggestion
Sometimes the UpdateSelenium script has a problem getting rid of an older chromedriver.exe, even if you think you have killed any running processes named `chromedriver.exe`. You might receive the error `"Cannot remove item. Access to the path ...\chromedriver.exe is denied"`, as in the screenshot above. In that case, you should follow the steps below:
```

> - Open the path shown in the error message in Explorer (e.g. C:\Users\tkolovos\Documents\WindowsPowerShell\Modules\Selenium\3.0.1\assemblies)

Reviewer: ```suggestion
- Open the path shown in the error message in Explorer (e.g. `C:\Users\tkolovos\Documents\WindowsPowerShell\Modules\Selenium\3.0.1\assemblies`)
```

```
product.<variant>.testplan.yaml
```
The optional variant can have values like: `g5`, `e5`, `gcc`, `pnp`, `spo`, `e#` etc. These `variant` test plans capture test cases unique to that tenant type and/or powershell type. 

## Functional Testing Usage ##

After setting up the prerequisites, define the Pester test container parameters a test execution utility script called "RunFunctionalTest.ps1" as shown below: 

```
$TestContainers = @()

$TestContainers += New-PesterContainer -Path "Testing/Functional/Products" -Data @{ TenantDomain = "MyG5tenant.onmicrosoft.com"; TenantDisplayName = "My G5 tenant"; ProductName = "aad"; M365Environment = "gcc" }

$PesterConfig = @{
	Run = @{
		Container = $TestContainers
	}
	Output = @{
		Verbosity = 'Detailed'
	}
}

$Config = New-PesterConfiguration -Hashtable $PesterConfig 

Invoke-Pester -Configuration $Config
```

> Copy the above utility script and save it into a file named "RunFunctionalTest.ps1" on your system. Save this file in a folder named FunctionalTesting that is at the same level in your directory tree as ScubaGear as shown below.

Reviewer: It should be noted that wherever one stores the utility script, it's important to run it script from the root of the ScubaGear directory, not from the directory where the utility script is stored.  Perhaps this could be added as a note.

Author: added.

$PesterConfig = @{
	Run = @{
		Container = $TestContainers
	}
	Output = @{
		Verbosity = 'Detailed'
	}
}

$Config = New-PesterConfiguration -Hashtable $PesterConfig 

Invoke-Pester -Configuration $Config

```

AAD functional testing for specific test cases: when developing or if you are testing pull requests you may need to execute just a single baseline policy in the respective YAML test plan file. To do that you will add Pester filter tags to the PesterConfig parameter. 

Use the following example execution script to run a test against AAD policy MS.AAD.2.1v1 using a service principal. 
```
$TestContainers = @()

$TestContainers += New-PesterContainer -Path "Testing/Functional/Products" -Data @{ Thumbprint = "D17F58D66DC5750EF78F25FF04CCE88DE730BCA6"; AppId = "1947b06c-46a9-4ff2-80c8-27261e58868b"; TenantDomain = "MyG5tenant.onmicrosoft.com"; TenantDisplayName = "My G5 tenant"; ProductName = "aad"; M365Environment = "gcc" }

$PesterConfig = @{
	Run = @{
		Container = $TestContainers
	}

>   Filter = @{

Reviewer: Indent properly.

Author: done

              Identity: Global
              AllowPSTNUsersToBypassLobby: false
          - Command: Set-CsTeamsMeetingPolicy
            Splat:
              AutoAdmittedUsers: EveryoneInCompany
              Identity: Global
        Postconditions: []
        ExpectedResult: true
```

## Nightly Functional Tests ##

A GitHub workflow is used to execute multiple product functional tests every night, and they are being run for multiple product-tenant combinations. Development teams should work with operational team to debug any issues found during nightly and/or weekly functional tests. The current list of tests can be found in a Slack canvas. 

## Troubleshooting ##

### Issues with installing Pester
If the Pester module is pre-installed and pre-loaded on your system, you may see following message when you attempt to install it:

```
WARNING: The version '5.5.0' of module 'Pester' is currently in use. Retry the operation after closing the
applications.
```

To Resolve above issue, try the following:
- Close the PowerShell session
- Open Task Manager to find any background instances of PowerShell and close them all 
- Uninstall all versions of Pester using following command
```

> Uninstall-Module -Name Pester -Force -AllVersions

Reviewer: Recommend indenting, like line 369 below, for consistency.

Author: done

  ```
  (GetInstalledModule -Name Pester).InstalledLocation
  ```
- Best practice is to install Pester and other modules as a non-admin user with either 'Current User' or 'AllUsers' scope. When modules are installed with 'AllUsers' scope, they are installed to the ``` $env:ProgramFiles\PowerShell\Modules``` location. The 'CurrentUser' scope installs modules to ``` $HOME\Documents\PowerShell\Modules``` location. 

### Iusses with Selenium
We have seen issues with Selenium where test orchestrator would not run and produce random errors. If you have seen issues where error messages does not match with any other items in this troubleshooting section, try uninstalling and re-installing Selenium using following commands:
```
uninstall-module Selenium
install-module Selenium
./Testing/Functional/SmokeTest/UpdateSelenium.ps1
```

### Chrome browser version issue ###

If the Chrome browser on your system has been updated since you setup the Selenium environment, you may receive an error message like below:

![Chrome browser version](/images/chrome-version.png)

> To resolve this, first kill any ChromeDriver.exe processes running by using the Task Manager. Note that ChromeDriver.exe processes can be hidden under the PowerShell app process and may not be visible in the main process list of Task Manager. Expand PowerShell app processes as shown in the image below to check for any hidden ChromeDriver processes - if you find any, kill them all.

Reviewer: ```suggestion
To resolve this, first kill any ChromeDriver.exe processes running by using the Task Manager. Note that ChromeDriver.exe processes can be hidden under the PowerShell app process and may not be visible in the main process list of Task Manager. Expand PowerShell app processes as shown in the image below to check for any hidden ChromeDriver processes, and if you find any, kill them all.
```

![Chrome browser processes](/images/chrome-process-kill2.png)

If you do not successfully kill that process you might receive an error message such as the one below.

![Chrome browser processes](/images/chrome-process-kill.png)

> Then execute the commands below to update the Selenium and Chrome driver.

Reviewer: ```suggestion
Execute the commands below to update the Selenium and Chrome driver.
```

```
.\Testing\Functional\SmokeTest\UpdateSelenium.ps1
```

> Sometimes the UpdateSelenium script has a problem getting rid of an older chromedriver.exe, even if you think you have killed any running processes named 'chromedriver.exe'. You might receive the error "Cannot remove item. Access to the path ...\chromedriver.exe is denied", as in the screenshot above. In that case you should follow the steps below:

Reviewer: ```suggestion
Sometimes the UpdateSelenium script has a problem getting rid of an older chromedriver.exe, even if you think you have killed any running processes named 'chromedriver.exe'. You might receive the error `Cannot remove item. Access to the path ...\chromedriver.exe is denied`, as shown in the screenshot above. In that case, you should follow the steps below:
```

> - Open the path shown in the error message in Explorer (e.g. 'C:\Users\tkolovos\Documents\WindowsPowerShell\Modules\Selenium\3.0.1\assemblies')

Reviewer: ```suggestion
- Open the path shown in the error message in Explorer (e.g. `C:\Users\username\Documents\WindowsPowerShell\Modules\Selenium\3.0.1\assemblies`)
```

- Rename the file to chromedriver.exe.bak
- Check the running processes in Task Manager for instances of a process named "chromedriver.exe.bak" and kill any that are shown
- Delete the file chromedriver.exe.bak from the drive
- Run the UpdateSelenium script again

### Service principal authentication issue ###

> If you are trying to run the test orchestrator as a service principal and your client certificate has NOT been uploaded into the Azure AD registered applications Certificates and Secrets page, you may receive an error similar to the one below (this is specific to the AAD product - other products may give slightly different errors). Contact the system admin to upload your certificate file.

Reviewer: ```suggestion
If you are trying to run the test orchestrator as a service principal and your client certificate has NOT been uploaded into the Azure AD registered applications Certificates and Secrets page, you may receive an error similar to the one below (this is specific to the AAD product; other products may give slightly different errors). Contact the system admin to upload your certificate file.
```
